### PR TITLE
Fix color processing steps

### LINKED ISSUE #760: Color processing steps are incorrect
We currently rely on [[CSS-SYNTAX-3]] "parse a component value", but that's an internal algorithm that's participating in a larger parser... it also skips of preconditions that the parser requires. 

We've [asked the CSS working group for guidance](https://lists.w3.org/Archives/Public/www-style/2019Apr/0014.html).

### PR DESCRIPTION
**Pull request by christianliebel**
Closes #760

This change (choose one):

* [ ] Breaks existing normative behavior (please add label "breaking")
* [ ] Adds new normative requirements
* [ ] Adds new normative recommendations or optional items
* [x] Makes only editorial changes (only changes informative sections, or
  changes normative sections without changing behavior)
* [ ] Is a "chore" (metadata, formatting, fixing warnings, etc).

Commit message:

Use the official extension point to parse a CSS color.

***
<a href="https://pr-preview.s3.amazonaws.com/christianliebel/manifest/pull/878.html" title="Last updated on Jun 2, 2020, 10:50 PM UTC (2ffa429)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3c/manifest/878/6bd6ca6...christianliebel:2ffa429.html" title="Last updated on Jun 2, 2020, 10:50 PM UTC (2ffa429)">Diff</a>

### CODE REVIEW

File: index.html [hunk 1]
```diff
@@ -1801,26 +1797,9 @@ <h3>
...
-              </li>
-              <li>Return <code>undefined</code>.
-              </li>
-            </ol>
-          </li>
-          <li>Return <var>color</var>.
-          </li>
-        </ol>
+        <div>
+          TODO
```
Comment (kenchris) [inline index.html:1801]: ??? :) Not ready for review

Comment (christianliebel) [reply index.html:1801]: Yep, this is WIP. 😇

**Comment by christianliebel** [comment]
@tabatkins Thanks for your help and support over at https://github.com/w3c/manifest/issues/760. Could you please check whether I have correctly included your suggestions here? Is it okay to reference CSS Color 4 for color conversions?

**Comment by christianliebel** [comment]
@marcoscaceres I'm facing some ReSpec issues here:
* `[=CSS/parsing=]` does not seem to be working correctly
  * warning: `Couldn't find a match for "css-syntax" at: 1.`
  * error: `Bad reference: [css-syntax] (appears 0 times)`
* ReSpec doesn’t seem to like the angle brackets of the `<color>` type.
  * error: `ReSpec doesn't support expanding this kind of reference. at: 1.`

Could you help me?

**Comment by christianliebel** [comment]
In general:
- As both `theme_color` and `background_color` will have to use the same algorithm, I was wondering if it makes sense to abstract this algorithm. However, I'm not sure where the abstracted version should go (section-wise).
- Should we clarify transparency handling here (e.g., the user agent MAY ignore the alpha value)?

**Comment by tabatkins** [comment]
Ah, you're editting a ReSpec spec. I can't help you with those errors, unfortunately. Most of the W3C edits their specs with Bikeshed, where my suggestions would work.

**Comment by tabatkins** [comment]
Otherwise, tho, the text looks good.

File: index.html [hunk 2]
```diff
@@ -1684,27 +1684,44 @@ <h3>
...
               <li>Return <code>undefined</code>.
               </li>
             </ol>
           </li>
-          <li>Return <var>color</var>.
+          <li>If |color| is an <a>sRGB</a> color, then return |color|.
+          </li>
+          <li>If |color| can be converted to <a>sRGB</a> using solely
+          information the user agent inherently knows, then convert it to
+          <a>sRGB</a> [[CSS-COLOR-4]], and return the result. Otherwise:
```
Comment (marcoscaceres) [inline index.html:1705]:
```suggestion
          <a>sRGB</a>, and return the result. Otherwise:
```

File: index.html [hunk 3]
```diff
@@ -3414,11 +3414,23 @@ <h2>
         expected.
       </p>
       <ul class="index" data-sort="">
+        <li>[[CSS-COLOR-4]] defines the following terms:
```
Comment (marcoscaceres) [inline index.html:3417]: Please delete these...

File: index.html [hunk 4]
```diff
@@ -3414,11 +3414,23 @@ <h2>
...
+              <a data-cite="CSS-COLOR-4#sRGB"><dfn>sRGB</dfn></a>
+            </li>
+          </ul>
+        </li>
         <li>[[CSS-SYNTAX-3]] defines the following terms:
           <ul>
             <li>
-              <a data-cite="CSS-SYNTAX-3#parse-component-value"><dfn>parse a
-              component value</dfn></a>
+              <a data-cite=
```
Comment (marcoscaceres) [inline index.html:3431]: Please delete this too

File: index.html [hunk 5]
```diff
@@ -1684,27 +1684,44 @@ <h3>
...
-          <li>Let <var>potential color</var> be the result of running
-          [[CSS-SYNTAX-3]]'s <a>parse a component value</a> algorithm with
-          <var>theme color</var> as input. If parsing returns a syntax error,
-          return <code>undefined</code>.
-          </li>
-          <li>Let <var>color</var> be the result of attempting to parse
-          <var>potential color</var> as a CSS color, as per [[CSS-SYNTAX-3]].
-          If parsing fails:
+          <li>Let |color| be the result of [=CSS/parsing=] |theme color| as a
+          CSS <a>&lt;color&gt;</a>. [[CSS-SYNTAX-3]]
```
Comment (marcoscaceres) [inline index.html:1689]: Note, "CSS/parsing" and "&lt;color&gt;" are not things in the exported terms database - at least that we can find at the moment via https://respec.org/xref... @sidvishnoi can check why they are not there.  

@christianliebel, when there is a linking error, ReSpec tries to tell you how to fix it at https://respec.org/xref. I can't see `sRGB` at https://respec.org/xref either.

Comment (marcoscaceres) [reply index.html:1689]: Found it: [=CSS/parse=] should work... 

```suggestion
          <li>Let |color| be the result of [=CSS/parse=] |theme color| as a
          CSS &lt;color&gt;.
```

File: index.html [hunk 6]
```diff
@@ -3414,11 +3414,23 @@ <h2>
...
+          </ul>
+        </li>
         <li>[[CSS-SYNTAX-3]] defines the following terms:
           <ul>
             <li>
-              <a data-cite="CSS-SYNTAX-3#parse-component-value"><dfn>parse a
-              component value</dfn></a>
+              <a data-cite=
+              "CSS-SYNTAX-3#css-parse-something-according-to-a-css-grammar"><dfn>
+              parse something according to a CSS grammar</dfn></a>
```
Comment (marcoscaceres) [inline index.html:3433]:
```suggestion
```

File: index.html (hunk 5 above)
Comment (sidvishnoi) [reply index.html]: `[=CSS/parsing=]` works, but need to add an alias for css-parsing in `respecConfig.localBiblio` until https://github.com/tobie/specref/pull/564 is merged.
``` js
localBiblio: {
  "css-parsing": { aliasOf: "css-parsing-3" },
}
```

File: index.html [hunk 7]
```diff
@@ -1684,27 +1684,44 @@ <h3>
...
+              </li>
+              <li>Return <code>undefined</code>.
+              </li>
+            </ol>
           </li>
         </ol>
+        <p class="note">
+          Only <a>sRGB</a> colors, and colors the user agent can convert to
+          <a>sRGB</a> without any outside knowledge, are supported. For
+          example, <code>lab(…)</code> or <code>color(display-p3, …)</code> can
```
Comment (marcoscaceres) [inline index.html:1719]:
```suggestion
          example, `lab(…)` or `color(display-p3, …)` can
```

File: index.html [hunk 8]
```diff
@@ -1684,27 +1684,44 @@ <h3>
           takes a <a>USVString</a> <var>theme color</var> as an argument. This
           algorithm returns a <code>USVString?</code>.
         </p>
-        <div class="issue" data-number="760"></div>
         <ol>
```
Comment (marcoscaceres) [inline index.html:1687]:
```suggestion
        <ol data-cite="CSS-SYNTAX-3 CSS-COLOR-4">
```

File: index.html [hunk 9]
```diff
@@ -1684,27 +1684,44 @@ <h3>
...
+              </li>
+            </ol>
           </li>
         </ol>
+        <p class="note">
+          Only <a>sRGB</a> colors, and colors the user agent can convert to
+          <a>sRGB</a> without any outside knowledge, are supported. For
+          example, <code>lab(…)</code> or <code>color(display-p3, …)</code> can
+          be converted to <a>sRGB</a> without outside knowledge, but
+          <code>color(--custom-profile, …)</code> would require finding a
```
Comment (marcoscaceres) [inline index.html:1721]:
```suggestion
          `color(--custom-profile, …)` would require finding a
```

**Comment by kenchris** [comment]
> In general:
> 
> * As both `theme_color` and `background_color` will have to use the same algorithm, I was wondering if it makes sense to abstract this algorithm. However, I'm not sure where the abstracted version should go (section-wise).
> * Should we clarify transparency handling here (e.g., the user agent MAY ignore the alpha value)?

We already have substracted algos like the within-scope one.

image

Color parsing
===

To obtain a color... 


For transparency I guess you can make a non-nominative note, saying that User Agents are free to deal with transparency the way they see fit

File: index.html (hunk 5 above)
Comment (tabatkins) [reply index.html]: > Found it: [=CSS/parse=] should work...

Note that Bikeshed automatically handles common conjugations of the term, so you can write things like `[=CSS/parsing=]` and have it automatically resolve to `[=CSS/parse=]`; perhaps ReSpec should do the same?

Comment (sidvishnoi) [reply index.html]: @tabatkins `[=CSS/parsing=]` also works in ReSpec, the problem was with Specref API
See https://github.com/w3c/manifest/pull/878#discussion_r432331678

**Comment by christianliebel** [comment]
Should the output be canonicalized maybe?

File: index.html [hunk 10]
```diff
@@ -1195,6 +1195,67 @@ <h3>
...
+        <p class="note" title="Supported colors">
+          Only [=sRGB=] colors, and colors the user agent can convert to
+          [=sRGB=] without any outside knowledge, are supported. For example,
+          `lab(…)` or `color(display-p3, …)` can be converted to [=sRGB=]
+          without outside knowledge, but `color(--custom-profile, …)` would
+          require finding a matching "@color-profile" rule which cannot be
+          specified in the manifest.
+        </p>
+        <p class="note" title="Transparent colors">
+          User agents can decide to override the color's [=alpha component=]
```
Comment (marcoscaceres) [inline index.html:1253]: Move to algorithm as a MAY.

File: index.html [hunk 11]
```diff
@@ -1195,6 +1195,67 @@ <h3>
...
+    <section>
+      <h2>
+        Color parsing
+      </h2>
+      <p>
+        This section defines an algorithm for <a>processing a color member</a>.
+      </p>
+      <section data-cite="CSS-COLOR-4">
+        <h3>
+          processing a color member
```
Comment (marcoscaceres) [inline index.html:1207]:
```suggestion
          Processing a color member
```

File: index.html [hunk 12]
```diff
@@ -1195,6 +1195,67 @@ <h3>
         <div class="issue" data-number="446"></div>
       </section>
     </section>
+    <section>
+      <h2>
```
Comment (marcoscaceres) [inline index.html:1199]: combine the two sections

File: index.html [hunk 13]
```diff
@@ -1195,6 +1195,59 @@ <h3>
...
     </section>
+    <section data-cite="CSS-COLOR-4">
+      <h2>
+        Color parsing
+      </h2>
+      <p>
+        The steps for <dfn>processing a color member</dfn> are given by the
+        following algorithm. The algorithm takes a <a>USVString</a>
+        <var>potential color</var> as an argument. This algorithm returns a
+        `USVString?`.
```
Comment (marcoscaceres) [inline index.html:1206]:
```suggestion
        `USVString` or `undefined`.
```

File: index.html [hunk 14]
```diff
@@ -1195,6 +1195,59 @@ <h3>
...
+              <a>Issue a developer warning</a> that |potential color| is not a
+              supported color value.
+            </li>
+            <li>Return `undefined`.
+            </li>
+          </ol>
+        </li>
+      </ol>
+      <p>
+        The user agent MAY override the resulting color's [=alpha component=]
```
Comment (marcoscaceres) [inline index.html:1238]: This should be moved into the algorithm also.

Comment (marcoscaceres) [reply index.html:1238]: Actually, no... sorry... this should move to theme color.

**Pull request merged**